blog/en/blog/2022/09/28/rust-loves-apisix.md

```diff
@@ -0,0 +1,198 @@
+
+The specification is available on [GitHub](https://github.com/proxy-wasm/spec).
+
+* Developers can create SDK for their tech stack
+* Proxy and API Gateway providers can integrate `proxy-wasm` in their product
+
+## Apache APISIX and proxy-wasm
+
+The Apache APISIX project decided to integrate `proxy-wasm` into the product to benefit from the standardization effort. It also allows end-users to start with Envoy, or any other `proxy-wasm`-compatible reverse proxy, to migrate to Apache APISIX when necessary.
+
+APISIX doesn't implement `proxy-wasm` but integrates [wasm-nginx-module](https://github.com/api7/wasm-nginx-module). It's an Apache v2-licensed project provided by [api7.ai](https://api7.ai/), one of the main contributors to Apache APISIX. As its name implies, integration is done at the NGINX level.
+
+![Apache APISIX and WebAssemby architecture overview](https://api7.ai/wp-content/uploads/2022/09/architecture-diagram.png)
+
+## Let's code!
+
+Now that we have explained how everything fits together, it's time to code.
+
+### Preparing Rust for WebAssembly
+
+Before developing the first line of code, we need to give Rust <abbr title="WebAssembly">WASM</abbr> compilation capabilities.
+
+```bash
+rustup target add wasm32-wasi
+```
+
+It allows the Rust compiler to output WASM code:
+
+```bash
+cargo build --target wasm32-wasi
+```
+
+The WASM code is found in:
+
+* `target/wasm32-wasi/debug/sample.wasm`
+* `target/wasm32-wasi/release/sample.wasm` (when compiled with the `--release` flag)
+
+### Setting up the project
+
+The setup of the project is pretty straightforward:
+
+```bash
+cargo new sample --lib
+```
+
+The command creates a `lib` project with the expected structure.
+
+### The code itself
+
+Let me first say that the available documentation is pretty sparse. For example, `proxy-wasm`'s is limited to the [methods' signature](https://github.com/proxy-wasm/spec/tree/master/abi-versions/vNEXT) (think JavaDocs). Rust SDK is sample-based. However, one can get some information from the [C++ SDK](https://github.com/proxy-wasm/proxy-wasm-cpp-sdk/blob/master/docs/wasm_filter.md).
+
+>WASM module is running in a stack-based virtual machine and its memory is isolated from the host environment. All interactions between host and WASM module are through functions and callbacks wrapped by context object.
+>
+>At bootstrap time, a root context is created. The root context has the same lifetime as the VM/runtime instance and acts as a target for any interactions which happen at initial setup. It is also used for interactions that outlive a request.
+>
+>At request time, a context with incremental is created for each stream. Stream context has the same lifetime as the stream itself and acts as a target for interactions that are local to that stream.
+
+The Rust code maps to the same abstractions.
+
+![Rust's 'structure diagram'](https://api7.ai/wp-content/uploads/2022/09/struct-diagram.png)
```

Comment:
ditto, Tencent CDN? cc @juzhiyuan 
```suggestion
![Rust's structure diagram](https://api7.ai/wp-content/uploads/2022/09/struct-diagram.png)
```

Reply:
There's no such thing as a structure diagram in UML, hence the quotes. I can remove them, I don't understand why.

Reply:
OK cool.